### PR TITLE
[TeamQuest] Define fields for the TeamQuest and TeamQuestStatus models

### PR DESCRIPTION
**Pull request by raztechs**

### CODE REVIEW

File: wouso/games/teamquest/models.py [hunk 1]
```diff
@@ -33,24 +40,22 @@ def __init__(self, *args, **kwargs):
...
-    head = models.OneToOneField('TeamQuestUser', null=True, blank=False)
-
-    @classmethod
-    def create(cls, head, name):
-        new_group = cls.objects.create(name=name, head=head)
-        new_group.users.add(head)
-        head.group = new_group
-        head.save()
-        return new_group
+    pass
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: Same here

File: wouso/games/teamquest/models.py [hunk 2]
```diff
@@ -2,19 +2,26 @@
...
-    def is_head(self):
-        if self.group is None:
-            return False
-        return self.group.head == self
+    pass
 
 
 class TeamQuest(models.Model):
-	pass
+    QUEST_TYPES = (
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: We should not have types for now. I think all the answers should be "In site text answers"

File: wouso/games/teamquest/models.py [hunk 3]
```diff
@@ -2,19 +2,26 @@
...
+        (TYPE_CLASSIC, 'In site text answers'),
+        (TYPE_CHECKER, 'In site answers, verified by checker'),
+        (TYPE_EXTERNAL, 'External levels and answers'),
+    )
+    start = models.DateTimeField()
+    end = models.DateTimeField()
+    title = models.CharField(default="", max_length=100)
+    questions = models.ManyToManyField(Question)
+    order = models.CharField(max_length=1000, default="", blank=True)
+    type = models.IntegerField(default=TYPE_CLASSIC, choices=QUEST_TYPES)
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: No type

File: wouso/games/teamquest/models.py [hunk 4]
```diff
@@ -2,19 +2,26 @@
...
+    QUEST_TYPES = (
+        (TYPE_CLASSIC, 'In site text answers'),
+        (TYPE_CHECKER, 'In site answers, verified by checker'),
+        (TYPE_EXTERNAL, 'External levels and answers'),
+    )
+    start = models.DateTimeField()
+    end = models.DateTimeField()
+    title = models.CharField(default="", max_length=100)
+    questions = models.ManyToManyField(Question)
+    order = models.CharField(max_length=1000, default="", blank=True)
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: We are not going to have a sort feature for the team quest

**Comment by iulianR** [comment]
I think we need another model: TeamQuestLevel.
It should contain a questions field (one level can have multiple questions) and a bonus. The TeamQuest should have a connection to it: one quest can have multiple levels - a Foreign Key.

@razvand @razvan3895 What do you think?

File: wouso/games/teamquest/models.py (hunk 4 above)
Comment (raztechs) [inline wouso/games/teamquest/models.py]: Ok. Will make the changes

**Comment by razvand** [comment]
@iulianR, if this model is required, we should take into account reducing the number of links between models.

An option would be for the TeamQuestLevel to have a foreign key to the TeamQuest. It may be better than having an array of levels linked from the TeamQuest model.

**Comment by iulianR** [comment]
That's what I had in mind, definitely a ForeignKey.

**Comment by razvand** [comment]
@iulianR, your proposal, from what I understand, is having an array of links to TeamQuestLevel instances from the TeamQuest. I was thinking of creating links from the TeamQuestLevel instances to the TeamQuest.

**Comment by iulianR** [comment]
I see it like this:
TeamQuestLevel:
- quest = ForeignKey (TeamQuest, related_name='levels')
- bonus = IntegerField

L\
L  ----> TeamQuest
L/

We are probably thinking at the same thing but I just phrased it badly.

File: wouso/games/teamquest/models.py [hunk 5]
```diff
@@ -14,7 +15,11 @@ def is_head(self):
 
 
 class TeamQuest(models.Model):
-	pass
+    start = models.DateTimeField()
+    end = models.DateTimeField()
+    title = models.CharField(default="", max_length=100)
+    questions = models.ManyToManyField(Question)
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: The questions are now part of the level (TeamQuestLevel). We don't need this anymore.

File: wouso/games/teamquest/models.py [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 from wouso.core.game.models import Game
 from wouso.core.user.models import Player, PlayerGroup
+from wouso.core.qpool.models import Question
```
Comment (iulianR) [inline wouso/games/teamquest/models.py]: This should be removed as it's needed for the TeamQuestLevel (which is not related to the current pull request)

File: wouso/games/teamquest/models.py [hunk 7]
```diff
@@ -75,7 +79,13 @@ def __unicode__(self):
 
 
 class TeamQuestStatus(models.Model):
-    pass
+    group = models.ForeignKey('TeamQuestGroup', null=True, blank=True, related_name='quests')
+    quest = models.ForeignKey('TeamQuest', null=True, blank=True, related_name='participants')
+    current_level = models.IntegerField(default=0, blank=True)
+    started_time = models.DateTimeField(default=datetime.datetime.now, blank=True, null=True)
+    finished_time = models.DateTimeField(default=None, blank=True, null=True)
+    finished = models.BooleanField(default=False, blank=True)
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Remove `finished`. We will check if `finished_time` is `None` if we want to know whether a quest has been completed.

File: wouso/games/teamquest/models.py [hunk 8]
```diff
@@ -75,7 +79,13 @@ def __unicode__(self):
 
 
 class TeamQuestStatus(models.Model):
-    pass
+    group = models.ForeignKey('TeamQuestGroup', null=True, blank=True, related_name='quests')
+    quest = models.ForeignKey('TeamQuest', null=True, blank=True, related_name='participants')
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Remove `related_name` for the moment.

File: wouso/games/teamquest/models.py [hunk 9]
```diff
@@ -75,7 +79,13 @@ def __unicode__(self):
...
+    current_level = models.IntegerField(default=0, blank=True)
+    started_time = models.DateTimeField(default=datetime.datetime.now, blank=True, null=True)
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Rename to `time_started`.

File: wouso/games/teamquest/models.py [hunk 10]
```diff
@@ -75,7 +79,13 @@ def __unicode__(self):
...
+    finished_time = models.DateTimeField(default=None, blank=True, null=True)
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Rename to `time_finished`.

File: wouso/games/teamquest/models.py [hunk 11]
```diff
@@ -14,7 +15,11 @@ def is_head(self):
 
 
 class TeamQuest(models.Model):
-	pass
+    start = models.DateTimeField()
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Add `null`, `blank` and, if required, `default` parameter. Rename to `start_time`.

File: wouso/games/teamquest/models.py [hunk 12]
```diff
@@ -14,7 +15,11 @@ def is_head(self):
...
+    end = models.DateTimeField()
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Add `null`, `blank` and, if required, `default` parameter. Rename to `end_time`.

File: wouso/games/teamquest/models.py [hunk 13]
```diff
@@ -14,7 +15,11 @@ def is_head(self):
...
+    title = models.CharField(default="", max_length=100)
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Add `null` and `blank` parameters.

File: wouso/games/teamquest/models.py [hunk 14]
```diff
@@ -75,7 +79,13 @@ def __unicode__(self):
 
 
 class TeamQuestStatus(models.Model):
-    pass
+    group = models.ForeignKey('TeamQuestGroup', null=True, blank=True, related_name='quests')
+    quest = models.ForeignKey('TeamQuest', null=True, blank=True, related_name='participants')
+    current_level = models.IntegerField(default=0, blank=True)
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Add `null` parameter.

File: wouso/games/teamquest/models.py [hunk 15]
```diff
@@ -14,7 +14,10 @@ def is_head(self):
 
 
 class TeamQuest(models.Model):
-	pass
+    start_time = models.DateTimeField(default=None, blank=True, null=True)
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: `null` and `blank` should be `False`, there should be no `default`. One should provide explicit times when creating a quest.

File: wouso/games/teamquest/models.py [hunk 16]
```diff
@@ -14,7 +14,10 @@ def is_head(self):
...
+    end_time = models.DateTimeField(default=None, blank=True, null=True)
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: `null` and `blank` should be `False`, there should be no `default`. One should provide explicit times when creating a quest.

File: wouso/games/teamquest/models.py [hunk 17]
```diff
@@ -14,7 +14,10 @@ def is_head(self):
...
+    title = models.CharField(default="", max_length=100, blank=True, null=True)
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Similar to the above, I would be explicit in providing a name for the quest. So no default, `blank` and `null` should pe `False`.

File: wouso/games/teamquest/models.py [hunk 18]
```diff
@@ -75,7 +77,12 @@ def __unicode__(self):
 
 
 class TeamQuestStatus(models.Model):
-    pass
+    group = models.ForeignKey('TeamQuestGroup', null=True, blank=True)
+    quest = models.ForeignKey('TeamQuest', null=True, blank=True)
+    current_level = models.IntegerField(default=0, blank=True, null=True)
+    time_started = models.DateTimeField(default=datetime.datetime.now, blank=True, null=True)
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Remove `blank` and `null`. `time_started` should be explicit.

File: wouso/games/teamquest/models.py [hunk 19]
```diff
@@ -75,7 +77,12 @@ def __unicode__(self):
 
 
 class TeamQuestStatus(models.Model):
-    pass
+    group = models.ForeignKey('TeamQuestGroup', null=True, blank=True)
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Remove `null` and `blank`.

File: wouso/games/teamquest/models.py [hunk 20]
```diff
@@ -75,7 +77,12 @@ def __unicode__(self):
...
+    quest = models.ForeignKey('TeamQuest', null=True, blank=True)
```
Comment (razvand) [inline wouso/games/teamquest/models.py]: Remove `null` and `blank`

**Comment by razvand** [comment]
All OK from my part. @iulianR, you can merge this pull request.

**Pull request merged**